test/run_test.py

```diff
@@ -339,19 +339,20 @@ def __contains__(self, item):
 
 
 if dist.is_available():
+    num_gpus = torch.cuda.device_count()
     DISTRIBUTED_TESTS_CONFIG["test"] = {"WORLD_SIZE": "1"}
     if not TEST_WITH_ROCM and dist.is_mpi_available():
         DISTRIBUTED_TESTS_CONFIG["mpi"] = {
             "WORLD_SIZE": "3",
         }
-    if dist.is_nccl_available():
+    if dist.is_nccl_available() and num_gpus > 0:
         DISTRIBUTED_TESTS_CONFIG["nccl"] = {
-            "WORLD_SIZE": f"{torch.cuda.device_count()}",
+            "WORLD_SIZE": f"{num_gpus}",
         }
-    if dist.is_gloo_available():
+    if dist.is_gloo_available() and num_gpus > 0:
         DISTRIBUTED_TESTS_CONFIG["gloo"] = {
             # TODO: retire testing gloo with CUDA
-            "WORLD_SIZE": f"{torch.cuda.device_count()}",
+            "WORLD_SIZE": f"{num_gpus}",
```

Comment:
As you mentioned, maybe we can remove the `num_gpus > 0` condition for Gloo, and set `WOLRD_SIZE` to `min(4, num_gpus)` ?

Reply:
I'm not sure about the "4". And `min` would still yield zero for missing GPUs which was the reason for this PR.

The previous world size was 3 and still is for MPI. The current (before this MR) intention is to use 1 rank for GPU.

So I'd use: `cuda_world_size = torch.cuda.device_count() or 3`. Or even `max(3, torch.cuda.device_count())`. That might skip some tests on single and dual-GPU systems but likely matches the previous and intended behavior.

What do you think?

Reply:
I just want to remove the `and num_gpus > 0` condition for Gloo, so that we would still run CPU tests on it.

Reply:
I see, especially with the TODO above it makes sense. I did the minimal change and use 3 processes when there are no GPUs, else the number of GPUs, i.e. the first suggestion